File: leetcode_problems/python/lc101_symmetric_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def check(
        self,
        l: TreeNode | None,
        r: TreeNode | None,
    ) -> bool:
        if l is None or r is None:
            return l is None and r is None
        if l.val != r.val:
            return False
        return self.check(l.left, r.right) and self.check(l.right, r.left)

    def isSymmetric(self, root: TreeNode | None) -> bool:
        if root is None:
            return False
        return self.check(root.left, root.right)


if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right = TreeNode(2)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)
    print(Solution().isSymmetric(root))

File: leetcode_problems/cpp/lc101_symmetric_tree.cpp
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool check(TreeNode *l, TreeNode *r) {
        if (l == nullptr && r == nullptr) {
            return true;
        } else if (l == nullptr || r == nullptr) {
            return false;
        } else if (l->val != r->val) {
            return false;
        } else {
            return check(l->left, r->right) && check(r->left, l->right);
        }
    }

    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }
        return check(root->left, root->right);
    }
};
